crates/polars-stream/src/nodes/parquet_source.rs

```diff
@@ -0,0 +1,1601 @@
+            let metadata_bytes = if bytes.len() - footer_size >= bytes.len() {
+                // Re-allocate to drop the excess bytes
+                MemSlice::from_vec(metadata_bytes.to_vec())
+            } else {
+                metadata_bytes
+            };
+
+            Ok((metadata_bytes, None))
+        }
+    }
+}
+
+fn get_row_group_byte_ranges(
+    row_group_metadata: &RowGroupMetaData,
+) -> impl Iterator<Item = std::ops::Range<usize>> + '_ {
+    let row_group_columns = row_group_metadata.columns();
+
+    row_group_columns.iter().map(|rg_col_metadata| {
+        let (offset, len) = rg_col_metadata.byte_range();
+        (offset as usize)..(offset + len) as usize
+    })
+}
+
+fn get_row_group_byte_ranges_for_projection<'a>(
+    row_group_metadata: &'a RowGroupMetaData,
+    columns: &'a [String],
+) -> impl Iterator<Item = std::ops::Range<usize>> + 'a {
+    let row_group_columns = row_group_metadata.columns();
+
+    row_group_columns.iter().filter_map(move |rg_col_metadata| {
+        for col_name in columns {
+            if &rg_col_metadata.descriptor().path_in_schema[0] == col_name {
+                let (offset, len) = rg_col_metadata.byte_range();
+                let range = (offset as usize)..((offset + len) as usize);
+                return Some(range);
+            }
+        }
+        None
+    })
+}
+
+/// Ensures that a parquet file has all the necessary columns for a projection with the correct
+/// dtype. There are no ordering requirements and extra columns are permitted.
+fn ensure_metadata_has_projected_fields(
+    projected_fields: &[polars_core::prelude::ArrowField],
+    metadata: &FileMetaData,
+) -> PolarsResult<()> {
+    let schema = polars_parquet::arrow::read::infer_schema(metadata)?;
+
+    // Note: We convert to Polars-native dtypes for timezone normalization.
+    let mut schema = schema
+        .fields
+        .into_iter()
+        .map(|x| {
+            let dtype = DataType::from_arrow(&x.data_type, true);
+            (x.name, dtype)
+        })
+        .collect::<PlHashMap<String, DataType>>();
+
+    for field in projected_fields {
+        let Some(dtype) = schema.remove(&field.name) else {
+            polars_bail!(SchemaMismatch: "did not find column: {}", field.name)
+        };
+
+        let expected_dtype = DataType::from_arrow(&field.data_type, true);
+
+        if dtype != expected_dtype {
+            polars_bail!(SchemaMismatch: "data type mismatch for column {}: found: {}, expected: {}",
+                &field.name, dtype, expected_dtype
+            )
+        }
+    }
+
+    Ok(())
+}
+
+async fn par_filter_df(df: &mut DataFrame, mask: &BooleanChunked) -> PolarsResult<()> {
+    let columns = unsafe { df.get_columns_mut() };
+
+    let handles = columns
+        .drain(..)
+        .map(|s| {
+            let mask = mask.clone();
+            async_executor::spawn(TaskPriority::Low, async move { s.filter(&mask) })
+        })
+        .collect::<Vec<_>>();
```

Comment:
@orlp I suspect what this `async fn par_filter_df` does might also not be desirable?

Reply:
We should already be parallel at this point, no?

Reply:
I added it as it's also what we do in the current batched Parquet reader. I did some benchmarking and it seems that this largely depends on the row group size - 

Using a file with 51 columns and 50M rows:

With a row group size of `~250K`:
```python
[ParquetSource]: Config { num_pipelines: 8, metadata_prefetch_size: 16, metadata_decode_ahead_size: 8, row_group_prefetch_size: 128, use_par_filter: false, use_par_decode: false }
python .env/y.py 2> /dev/null > /dev/null  34.89s user 36.52s system 733% cpu 9.741 total
python .env/y.py 2> /dev/null > /dev/null  32.80s user 39.32s system 725% cpu 9.944 total
python .env/y.py 2> /dev/null > /dev/null  35.90s user 38.21s system 726% cpu 10.204 total
python .env/y.py 2> /dev/null > /dev/null  35.34s user 36.53s system 722% cpu 9.944 total
python .env/y.py 2> /dev/null > /dev/null  34.09s user 37.17s system 728% cpu 9.776 total
[ParquetSource]: Config { num_pipelines: 8, metadata_prefetch_size: 16, metadata_decode_ahead_size: 8, row_group_prefetch_size: 128, use_par_filter: true, use_par_decode: false }
python .env/y.py 2> /dev/null > /dev/null  33.47s user 34.31s system 664% cpu 10.205 total
python .env/y.py 2> /dev/null > /dev/null  33.07s user 33.54s system 653% cpu 10.200 total
python .env/y.py 2> /dev/null > /dev/null  33.42s user 34.81s system 661% cpu 10.321 total
python .env/y.py 2> /dev/null > /dev/null  32.64s user 34.22s system 651% cpu 10.263 total
python .env/y.py 2> /dev/null > /dev/null  33.58s user 34.14s system 647% cpu 10.457 total

```

With a row group size of `~1M`:
```python
[ParquetSource]: Config { num_pipelines: 8, metadata_prefetch_size: 16, metadata_decode_ahead_size: 8, row_group_prefetch_size: 128, use_par_filter: false, use_par_decode: false }
python .env/y.py 2> /dev/null > /dev/null  32.54s user 62.53s system 697% cpu 13.635 total
python .env/y.py 2> /dev/null > /dev/null  37.92s user 56.94s system 689% cpu 13.753 total
python .env/y.py 2> /dev/null > /dev/null  40.86s user 55.84s system 676% cpu 14.298 total
python .env/y.py 2> /dev/null > /dev/null  41.94s user 59.19s system 699% cpu 14.451 total
python .env/y.py 2> /dev/null > /dev/null  42.94s user 56.38s system 721% cpu 13.768 total
[ParquetSource]: Config { num_pipelines: 8, metadata_prefetch_size: 16, metadata_decode_ahead_size: 8, row_group_prefetch_size: 128, use_par_filter: true, use_par_decode: false }
python .env/y.py 2> /dev/null > /dev/null  38.92s user 43.39s system 717% cpu 11.474 total
python .env/y.py 2> /dev/null > /dev/null  37.68s user 42.60s system 692% cpu 11.592 total
python .env/y.py 2> /dev/null > /dev/null  37.30s user 42.36s system 711% cpu 11.194 total
python .env/y.py 2> /dev/null > /dev/null  35.82s user 41.17s system 704% cpu 10.922 total
python .env/y.py 2> /dev/null > /dev/null  33.47s user 39.55s system 703% cpu 10.383 total
```

I think it should be ok to keep? We drop a little bit (~7%) for small row groups, but for larger ones we get (~25%) faster. Or if we want I could also dispatch between the 2 depending on the row group size

Reply:
Would it be possible to split up the larger row groups at an earlier step, so the vertical parallelism is fast again?

Reply:
I think that's possible - will work on it

Reply:
I've added functionality that splits large row groups and left a new comment below